socorro/stage_submitter/submitter.py

```diff
@@ -0,0 +1,455 @@
+                # so we need to convert that back; however, it's not clear this works
+                # in _all_ cases
+                if "dump" in dumps.keys():
+                    dumps["upload_file_minidump"] = dumps["dump"]
+                    del dumps["dump"]
+
+                payload_type = get_payload_type(raw_crash)
+                payload_compressed = get_payload_compressed(raw_crash)
+
+                # Get the metadata.user_agent if there is one, or use default agent
+                user_agent = (
+                    raw_crash.get("metadata", {}).get("user_agent")
+                    or DEFAULT_USER_AGENT
+                )
+
+                # Remove keys created by the collector from the raw crash
+                raw_crash = remove_collector_keys(raw_crash)
+
+                # Assemble payload and headers
+                payload, headers = multipart_encode(
+                    raw_crash=raw_crash,
+                    dumps=dumps,
+                    payload_type=payload_type,
+                    payload_compressed=payload_compressed,
+                )
+
+                # Set the User-Agent header so the collector captures this in the metadata
+                headers["User-Agent"] = user_agent
+
+                # Post to all destinations
+                for destination in destinations:
+                    try:
+                        # POST crash to new environment
+                        # FIXME(willkg): retryable post
+                        requests.post(destination.url, headers=headers, data=payload)
+
+                    except Exception:
+                        METRICS.incr("unknown_submit_error", value=1)
+                        self.logger.exception(
+                            "Error: http post failed for unknown reason: %s", crash_id
+                        )
+
+    def run_loop(self):
+        """Run cache manager in a loop."""
+        crashes = self.source_iterator()
+        while True:
+            crash = next(crashes)
+            crash_id = crash.crash_id
+            finished_func = crash.finished_func
+
+            try:
+                self.process(crash)
+            except Exception:
+                METRICS.incr("unknown_process_error", value=1)
+                self.logger.exception(
+                    "error: processing failed for unknown reason: %s", crash_id
+                )
+            finally:
+                try:
+                    finished_func()
+                except Exception:
+                    METRICS.incr("unknown_finished_func_error", value=1)
+                    self.logger.exception(
+                        "error: finished_func failed for unknown reason: %s", crash_id
+                    )
+
+        self.shutdown()
+
+    def run_once(self):
+        """Runs a nonblocking event generator once."""
+        crashes = self.source_iterator()
+        crash = next(crashes)
+        self.process(crash)
```

Comment:
The tests should run finish_func.

Reply:
Also, we can rework this function so it's not problematic repeated code that undermines our confidence.